import { Injectable, Inject, ConflictException, BadRequestException, InternalServerErrorException } from '@nestjs/common';
import { CreateProductDto } from 'src/dtos/product/dto/create-product.dto';
import { ProductDto } from 'src/dtos/product/dto/product.dto';
import { ProductRepository } from '../interfaces/product.interface';

@Injectable()
export class CreateProductService {
  constructor(
    @Inject('ProductRepository') 
    private readonly repository: ProductRepository,
  ) {}

  async execute(data: CreateProductDto): Promise<ProductDto> {
    try {
      // ðŸ”¹ Verifica se os campos obrigatÃ³rios estÃ£o preenchidos

      // ðŸ”¹ Verifica se registros Ãºnicos jÃ¡ existem no banco

      // âœ… Criar o registro
      return await this.repository.create(data);
    } catch (error) {
      console.error("Erro ao criar Product:", error.message);
      
      // ðŸ”¹ RelanÃ§a erros conhecidos diretamente
      if (error instanceof BadRequestException || error instanceof ConflictException) {
        throw error;
      }

      // ðŸ”¹ Erro inesperado, relanÃ§ado como erro interno do servidor
      throw new InternalServerErrorException("Erro ao criar Product.");
    }
  }
}
